### PR TITLE
add: callback trait proposed in rfc#0012

### PR DESCRIPTION
**Pull request by Velfi**
## Motivation and Context

Implements the [Body Callback RFC](https://github.com/awslabs/smithy-rs/blob/main/design/src/rfcs/rfc0012_body_callback_apis.md)

## Description

add: callback setting API to ByteStream
add: callback setting API to SdkBody
add: callback tests for streaming and buffered data
add: private callback fns for calculating checksums
udpate: pub methods on private struct `Inner` to be private
update: changelog
remove: Sync bound from SdkBody test "sdk_body_is_send()"
add: fn to merge `HeaderMap`s by appending
add: tests for new functionality

## Testing

ran existing tests and created new ones

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Velfi** [comment]
I'm putting this up for initial review. It has two TODOs that I could use some advising on. Currently, the checksum callbacks are private because I don't want people to start messing with them. This PR, once merged, will be followed by another PR that implements the codegen side of things and makes the callbacks public.

File: rust-runtime/aws-smithy-http/Cargo.toml [hunk 1]
```diff
@@ -29,16 +32,30 @@ hyper = "0.14"
...
 futures-core = "0.3.14"
 tokio = { version = "1.6", optional = true }
-tokio-util = { version = "0.6", optional = true}
+tokio-util = { version = "0.6", optional = true }
+
+# Checksum algos
+crc32fast = "1.3"
+crc32c = "0.6"
+sha1 = "0.10"
+sha2 = "0.10"
```
Comment (rcoh) [inline rust-runtime/aws-smithy-http/Cargo.toml:41]: I think ideally we'd keep this separate so we didn't compile all these deps when they weren't needed

File: rust-runtime/aws-smithy-http/src/callback.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+use http::{HeaderMap, HeaderValue};
+
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+/// A callback that, when inserted into a request body, will be called for corresponding lifecycle events.
+pub trait BodyCallback: Send {
+    /// This lifecycle function is called for each chunk **successfully** read. If an error occurs while reading a chunk,
+    /// this method will not be called. This method takes `&mut self` so that implementors may modify an implementing
+    /// struct/enum's internal state. Implementors may return an error.
+    fn update(&mut self, #[allow(unused_variables)] bytes: &[u8]) -> Result<(), BoxError> {
```
Comment (rcoh) [inline rust-runtime/aws-smithy-http/src/callback.rs:19]: just for visual noise, might be nicer to add `let _ = bytes;` in the body instead of the unused variables allow

File: rust-runtime/aws-smithy-http/src/callback/checksum.rs [hunk 3]
```diff
@@ -0,0 +1,260 @@
+/*
```
Comment (rcoh) [inline rust-runtime/aws-smithy-http/src/callback/checksum.rs:1]: we should probably keep these in another crate we can pull in when checksums are actually used

File: rust-runtime/aws-smithy-http/src/callback.rs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+        let body = ByteStream::from(body).collect().await.unwrap().into_bytes();
+        let body = std::str::from_utf8(&body).unwrap();
+
+        // Assert that the body that went in is the same as the body coming out.
+        assert_eq!(body, "test");
+
+        assert!(logs_contain("callback A was called"));
+        assert!(logs_contain("callback B was called"));
+        assert!(!logs_contain("callback was called"));
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-http/src/callback.rs:98]: What would have logged this?

File: rust-runtime/aws-smithy-http/src/header.rs [hunk 5]
```diff
@@ -295,16 +294,75 @@ pub fn quote_header_value<'a>(value: impl Into<Cow<'a, str>>) -> Cow<'a, str> {
...
+///     append_merge_header_maps(left_hand_side_headers, right_hand_side_headers);
+/// let merged_values: Vec<_> = merged_header_map
+///     .get_all(header_name.clone())
+///     .into_iter()
+///     .collect();
+///
+/// // Will print 'some_key: ["lhs value", "rhs value"]'
+/// println!("{}: {:?}", header_name.as_str(), merged_values);
+/// ```
+pub fn append_merge_header_maps(
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-http/src/header.rs:330]: This should be `pub(crate)`

File: rust-runtime/aws-smithy-http/src/header.rs [hunk 6]
```diff
@@ -560,4 +618,28 @@ mod test {
         assert_eq!("\"(\"", &quote_header_value("("));
         assert_eq!("\")\"", &quote_header_value(")"));
     }
+
+    #[test]
+    fn test_append_merge_header_maps() {
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-http/src/header.rs:623]: There should also be a test for multiple header values with the same header name in the same header map. Probably also a good idea to have tests with empty header maps.

File: rust-runtime/aws-smithy-checksum-callbacks/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,35 @@
+[package]
+name = "aws-smithy-checksum-callbacks"
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-checksum-callbacks/Cargo.toml:2]: Maybe just `aws-smithy-checksums`? The callback part seems like an implementation detail.

File: rust-runtime/aws-smithy-http/src/callback.rs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            .with_callback(Box::new(CallbackB));
+
+        let body = ByteStream::from(body).collect().await.unwrap().into_bytes();
+        let body = std::str::from_utf8(&body).unwrap();
+
+        // Assert that the body that went in is the same as the body coming out.
+        assert_eq!(body, "test");
+
+        assert!(logs_contain("callback A was called"));
+        assert!(logs_contain("callback B was called"));
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-http/src/callback.rs:97]: Do we care what order these were called in? I don't think the log testing will inform on that.

**Review by jdisanti — COMMENTED** [review]
Looks great! I just have some minor comments and bike shedding.

File: rust-runtime/aws-smithy-http/src/callback.rs (hunk 8 above)
Comment (Velfi) [reply rust-runtime/aws-smithy-http/src/callback.rs:97]: I think no. I'm uncomfortable with users depending on callbacks running in a specific order because it just seems like that would be a source of bugs/confusion. This test exists to show that all callbacks get called when setting more than one.

File: rust-runtime/aws-smithy-http/src/callback.rs (hunk 4 above)
Comment (Velfi) [reply rust-runtime/aws-smithy-http/src/callback.rs:98]: nothing, and I can't recall why I added it. I'll remove it.

File: rust-runtime/aws-smithy-checksums/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+        header_map.insert(key, value);
+
+        Ok(Some(header_map))
+    }
+}
+
+impl BodyCallback for Crc32callback {
+    fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
+        self.update(bytes)
+    }
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-checksums/src/lib.rs:51]: On all of these: why implement `update` and `trailers` in a separate impl block?

**Review by jdisanti — APPROVED** [review]
LGTM!

File: rust-runtime/aws-smithy-checksums/src/lib.rs (hunk 9 above)
Comment (Velfi) [reply rust-runtime/aws-smithy-checksums/src/lib.rs:51]: The original idea would be that it allows you to call these functions without needing to use the trait. This would happen within the "mutate request" section for non-streaming requests. For streaming requests they'd all be boxed trait objects.

I'm fine with making them only usable behind the trait if you like. Also, I don't think I've mentioned that these callbacks are going to be marked `pub` when I do the codegen half of things. I made them private to start so people wouldn't touch them.
